Fix pipeline TN93 output handling in true_append

stevenweaver opened the pull request:
## Summary
- Fixed issues with TN93 output file generation when true_append is executed as part of a pipeline
- Refactored file handling to use context managers consistently
- Uses temporary files to properly handle TN93 outputs
- Ensures only one header appears at the top of the file
- Fixed an issue in hivtrace.py with error handling

## Problem
When tn93 True Append is executed as a pipeline, some TN93 distances fail to appear in the output file, but this behavior doesn't happen when run as a standalone application. This was due to improper file handling and header management in the TN93 output generation.

## Test plan
Tested with the example test case in the hivtrace-secure-server using a previous run's file as input and verifying proper functioning (hi.sh test script).

niemasd committed:
Update true_append.py

stevenweaver committed:
Fix pipeline TN93 output handling in true_append

When executed as part of a pipeline, the true_append function was producing incorrect TN93 output files due to improper file handling. This PR:

1. Refactors file handling to use context managers consistently
2. Uses temporary files to properly handle TN93 outputs
3. Ensures only one header appears at the top of the file
4. Fixes an issue in hivtrace.py with error handling

niemasd commented:
@stevenweaver I see that you refactored the code to use temporary files (I had suggested this before I caught the unclosed output file), but I don't think that should be necessary after I adjusted the previous code to properly close the output file. I think we should stick with named pipes (rather than a temporary file) to avoid unnecessary disk I/O. In other words, I think just my initial commit in this PR + potentially the changes you made to the code that calls the True Append code should be sufficient. Thoughts?

niemasd committed:
Revert True Append to named pipe

In an attempted fix, @stevenweaver switched from a named pipe to a temporary file. However, I think the bug was elsewhere (I had forgotten to close the output file stream; I don't think the issue actually ended up having anything to do with the named pipe). I have reverted some of the changes to go back to the prior logic that used the named pipe.

niemasd commented:
@stevenweaver I pushed changes to the [`fix-true-append-pipeline` branch](https://github.com/veg/hivtrace/tree/fix-true-append-pipeline) that undo the temporary file stuff and instead reverts to the named pipe approach: I don't think the issue actually was because of named pipes, but rather, I think it was just because (1) I forgot to close the output distance file at the end, and (2) I forgot to redirect `tn93`'s standard output to the output distance file in one of the `tn93` calls (thanks for catching that!).

I kept the error-checking you added at the end (refactored to raise a `RuntimeError` instead of just printing the error message to the log). I also kept the refactors you made to the "copy old TN93 distances" code that made it more human-readable (I had all of the "copy this line" logic in a single convoluted `if` condition, which you broke up into more human-readable chunks).

Can you try running it on your test dataset and verify that it still works? If it does, I would prefer going with this version to avoid unnecessary disk I/O (as well as to remove the need to potentially manually clean up a temporary file)

niemasd commented:
Actually, I was able to get this fork of `hivtrace` running on the test dataset from my True Append repo. I was able to replicate a bug; I'll work on a fix and reply once it's ready to test

niemasd committed:
Manually flush output file after each step of tn93 True Append

niemasd commented:
@stevenweaver When I tested `tn93` True Append on some small example sequences ([NIEMA-1.TXT](https://github.com/user-attachments/files/20320232/NIEMA-1.TXT) and [NIEMA-2.TXT](https://github.com/user-attachments/files/20320234/NIEMA-2.TXT)), the header line ended up somewhere in the middle of the file, seemingly because of issues with not flushing the output file stream between the steps of the pipeline. I just pushed an update that explicitly flushes the output file stream between each step of the True Append process, which seems to have fixed it

When you get a chance, it would be great if you could try running the most recent version of this branch on the dataset that was giving you issues. Hopefully manually flushing the output file stream after each step of the process fixes the issue!

niemasd committed:
Renamed "HIV-TRACE True Append" to "TN93 True Append"

In anticipation of a future `cawlign` True Append as well as a future `DataQC` True Append, I renamed "HIV-TRACE True Append" to "TN93 True Append" since this code segment is just TN93. In the future, I can migrate the other components of True Append to this single `true_append.py` file

stevenweaver commented:
Dear @niemasd, 

Yes, I've verified that the latest commit works. 

Thanks so much,
Steven
